cloudconfig/machinecloudconfig.go

```diff
@@ -0,0 +1,291 @@
+func getMachineCloudCfgDirData(series string) (map[string]interface{}, error) {
+	dir, err := CloudInitCfgDir(series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot determine CloudInitCfgDir for the machine")
+	}
+	fileInfo, err := ioutil.ReadDir(dir)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot determine files in CloudInitCfgDir for the machine")
+	}
+	sortedFileInfos := sortableFileInfos(fileInfo)
+	sort.Sort(sortedFileInfos)
+	cloudInit := make(map[string]interface{})
+	for _, file := range fileInfo {
+		name := file.Name()
+		if !strings.HasSuffix(name, ".cfg") {
+			continue
+		}
+		data, err := ioutil.ReadFile(filepath.Join(dir, name))
+		if err != nil {
+			return nil, errors.Annotatef(err, "cannot read %q from machine", name)
+		}
+		cloudCfgData, err := unmarshallContainerCloudInit(data)
+		if err != nil {
+			return nil, errors.Annotatef(err, "cannot unmarshall %q from machine", name)
+		}
+		for k, v := range cloudCfgData {
+			cloudInit[k] = v
+		}
+	}
+	return cloudInit, nil
+}
+
+// getMachineVendorData returns a map of machine's cloud init vendor-data.txt.
+func getMachineVendorData(series string) (map[string]interface{}, error) {
+	// vendor-data.txt may or may not be compressed.
+	return getMachineData(series, "vendor-data.txt")
+}
+
+// MachineCloudInitDir is for testing purposes.
+var MachineCloudInitDir = paths.MachineCloudInitDir
+
+func getMachineData(series, file string) (map[string]interface{}, error) {
+	dir, err := MachineCloudInitDir(series)
+	if err != nil {
+		return nil, errors.Annotate(err, "cannot determine MachineCloudInitDir for the machine")
+	}
+	data, err := ioutil.ReadFile(filepath.Join(dir, file))
+	if err != nil {
+		return nil, errors.Annotatef(err, "cannot read %q from machine (%s)", file)
+	}
+
+	if len(data) == 0 {
+		// vendor-data.txt is sometimes empty
+		return nil, nil
+	}
+	// The userdata maybe be gzip'd, base64 encoded, both, or neither.
+	// If both, it's been gzip'd, then base64 encoded.
+	rawBuf, err := unmarshallContainerCloudInit(data)
+	if err == nil {
+		return rawBuf, nil
+	}
+	logger.Tracef("unmarshal of %q failed (%s), maybe it is compressed", file, err)
+
+	zippedData, err := utils.Gunzip(data)
+	if err == nil {
+		return unmarshallContainerCloudInit(zippedData)
+	}
+	logger.Tracef("Gunzip of %q failed (%s), maybe it is encoded", file, err)
+
+	decodedData, err := base64.StdEncoding.DecodeString(string(data))
+	if err == nil {
+		// it could still be gzip'd.
+		buf, err := unmarshallContainerCloudInit(decodedData)
+		if err == nil {
+			return buf, nil
+		}
+	}
+	logger.Tracef("Decoding of %q failed (%s), maybe it is encoded and gzipped", file, err)
+
+	decodedZippedBuf, err := utils.Gunzip(decodedData)
+	if err != nil {
+		// During testing, it was found that the trusty vendor-data.txt.i file
+		// can contain only the text "NONE", which doesn't unmarshall or decompress
+		// we don't want to fail in that case.
+		if series == "trusty" {
```

Comment:
we should check decodedData contents before assuming it's ok to return nil, nil

Reply:
decodedData contents were checked above, lines 143-150, and failed.